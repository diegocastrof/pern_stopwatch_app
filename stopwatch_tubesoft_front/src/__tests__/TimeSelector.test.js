import React from 'react';
import { shallow } from 'enzyme';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import TimeSelector from '../components/TimeSelector';

describe('Time Selector Component', () => {
	beforeEach(() => jest.useFakeTimers());
	afterEach(cleanup);

	// Component mounting test
	it('should render correctly', () => {
		const wrapper = shallow(
			<TimeSelector setFieldValue={() => {}} setIsDisabled={() => {}} />
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('timer should start at 00:00:00', () => {
		const { getByTestId } = render(
			<TimeSelector setFieldValue={() => {}} setIsDisabled={() => {}} />
		);
		expect(getByTestId('timer').textContent).toBe('00 : 00 : 00');
	});

	// DOM Show / hide elements test
	it('should hide start button & show pause button when running', () => {
		const { getByText, queryByText } = render(
			<TimeSelector setFieldValue={() => {}} setIsDisabled={() => {}} />
		);

		expect(queryByText('Pause')).toBeNull();

		fireEvent.click(getByText('Start'));
		expect(getByText('Pause')).toBeInTheDocument();
		expect(queryByText('Start')).toBeNull();

		fireEvent.click(getByText('Pause'));
		expect(getByText('Continue')).toBeInTheDocument();
		expect(queryByText('Pause')).toBeNull();
	});
	// Stopwatch functionality test
	it('should start timer on Start click event', () => {
		const { getByText, getByTestId } = render(
			<TimeSelector setFieldValue={() => {}} setIsDisabled={() => {}} />
		);

		fireEvent.click(getByText('Start'));
		act(() => {
			jest.advanceTimersByTime(5000);
		});
		expect(getByTestId('timer').textContent).toBe('00 : 00 : 05');
	});

	it('should add times if there is a pause between timetakes', () => {
		const { getByText, getByTestId } = render(
			<TimeSelector setFieldValue={() => {}} setIsDisabled={() => {}} />
		);

		// 1st Timetake
		fireEvent.click(getByText('Start'));
		act(() => {
			jest.advanceTimersByTime(5000);
		});
		fireEvent.click(getByText('Pause'));
		// 2nd Timetake
		fireEvent.click(getByText('Continue'));
		act(() => {
			jest.advanceTimersByTime(3000);
		});
		fireEvent.click(getByText('Pause'));

		expect(getByTestId('timer').textContent).toBe('00 : 00 : 08');
	});

	it('should should register when clock is paused after running', () => {
		const mockedSetFieldValue = jest.fn();
		const { getByText } = render(
			<TimeSelector
				setFieldValue={mockedSetFieldValue}
				setIsDisabled={() => {}}
			/>
		);

		fireEvent.click(getByText('Start'));
		act(() => {
			jest.advanceTimersByTime(5000);
		});
		fireEvent.click(getByText('Pause'));
		expect(mockedSetFieldValue).toBeCalledWith('time[time]', '00 : 00 : 05');
	});

	it('should unregister time when Reset', () => {
		const mockedSetFieldValue = jest.fn();
		const { getByText, getByTestId } = render(
			<TimeSelector
				setFieldValue={mockedSetFieldValue}
				setIsDisabled={() => {}}
			/>
		);

		fireEvent.click(getByText('Start'));
		act(() => {
			jest.advanceTimersByTime(5000);
		});
		fireEvent.click(getByText('Reset'));
		expect(getByTestId('timer').textContent).toBe('00 : 00 : 00');
		expect(mockedSetFieldValue).toBeCalledWith('time[time]', '');
	});

	it('should unregister time when Reset', () => {
		const mockedSetFieldValue = jest.fn();
		const { getByText, getByTestId } = render(
			<TimeSelector
				setFieldValue={mockedSetFieldValue}
				setIsDisabled={() => {}}
			/>
		);

		fireEvent.click(getByText('Start'));
		act(() => {
			jest.advanceTimersByTime(5000);
		});
		fireEvent.click(getByText('Reset'));
		expect(getByTestId('timer').textContent).toBe('00 : 00 : 00');
		expect(mockedSetFieldValue).toBeCalledWith('time[time]', '');
	});

	it('should disable submit button when a time is registered', () => {
		const mockedSetIsDisabled = jest.fn();
		const { getByText } = render(
			<TimeSelector
				setFieldValue={() => {}}
				setIsDisabled={mockedSetIsDisabled}
			/>
		);

		fireEvent.click(getByText('Start'));
		act(() => {
			jest.advanceTimersByTime(5000);
		});
		fireEvent.click(getByText('Pause'));
		expect(mockedSetIsDisabled).toBeCalledWith(false);
	});
});
