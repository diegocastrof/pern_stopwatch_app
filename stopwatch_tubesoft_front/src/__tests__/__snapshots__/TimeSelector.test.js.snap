// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time Selector Component should render correctly 1`] = `
<Fragment>
  <p
    className="text-center mt-5 lead font-weight-bold"
    data-testid="timer"
  >
    00 : 00 : 00
  </p>
  <Row
    className="mb-4"
    noGutters={false}
  >
    <Col
      className="d-flex justify-content-around"
    >
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Start
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Reset
      </Button>
    </Col>
  </Row>
</Fragment>
`;
